librz/il/definitions/bitvector.c

```diff
@@ -0,0 +1,1042 @@
+	bool mx = rz_il_bv_msb(x);
+	bool my = rz_il_bv_msb(y);
+
+	RzILBitVector neg_x, neg_y, tmp, ret;
+
+	if ((!mx) && (!my)) {
+		return rz_il_bv_div(x, y);
+	}
+
+	if ((mx) && (!my)) {
+		neg_x = rz_il_bv_neg(x);
+		tmp = rz_il_bv_div(neg_x, y);
+		ret = rz_il_bv_neg(tmp);
+
+		rz_il_bv_free(tmp);
+		rz_il_bv_free(neg_x);
+		return ret;
+	}
+
+	if ((!mx) && (my)) {
+		neg_y = rz_il_bv_neg(y);
+		tmp = rz_il_bv_div(x, neg_y);
+		ret = rz_il_bv_neg(tmp);
+
+		rz_il_bv_free(tmp);
+		rz_il_bv_free(neg_y);
+		return ret;
+	}
+
+	if (mx && my) {
+		neg_x = rz_il_bv_neg(x);
+		neg_y = rz_il_bv_neg(y);
+
+		ret = rz_il_bv_div(neg_x, neg_y);
+		rz_il_bv_free(neg_x);
+		rz_il_bv_free(neg_y);
+		return ret;
+	}
+
+	return NULL; // something wrong
+}
+
+/**
+ * Result of (x mod y) mod 2^length (signed algorithm)
+ *                            /
+ *                          | x % y : if not mx /\ not my
+ *                         | neg (neg x % y) if mx /\ not my
+ *           x smodulo y = <
+ *                          | neg (x % (neg y)) if not mx /\ my
+ *                          | neg (neg x % neg y) mod m if mx /\ my
+ *                          \
+ *
+ *           where mx = msb x  and my = msb y.
+ * \param x RzILBitVector, Operand
+ * \param y RzILBitVector, Operand
+ * \return ret RzILBitVector, point to the new bitvector
+ */
+RZ_API RzILBitVector rz_il_bv_smod(RzILBitVector x, RzILBitVector y) {
+	bool mx = rz_il_bv_msb(x);
+	bool my = rz_il_bv_msb(y);
+
+	RzILBitVector neg_x, neg_y, tmp, ret;
+
+	if ((!mx) && (!my)) {
+		return rz_il_bv_mod(x, y);
+	}
+
+	if ((mx) && (!my)) {
+		neg_x = rz_il_bv_neg(x);
+		tmp = rz_il_bv_mod(neg_x, y);
+		ret = rz_il_bv_neg(tmp);
+
+		rz_il_bv_free(tmp);
+		rz_il_bv_free(neg_x);
+		return ret;
+	}
+
+	if ((!mx) && (my)) {
+		neg_y = rz_il_bv_neg(y);
+		tmp = rz_il_bv_mod(x, neg_y);
+		ret = rz_il_bv_neg(tmp);
+
+		rz_il_bv_free(tmp);
+		rz_il_bv_free(neg_y);
+		return ret;
+	}
+
+	if (mx && my) {
+		neg_x = rz_il_bv_neg(x);
+		neg_y = rz_il_bv_neg(y);
+
+		tmp = rz_il_bv_mod(neg_x, neg_y);
+		ret = rz_il_bv_neg(tmp);
+		rz_il_bv_free(neg_x);
+		rz_il_bv_free(neg_y);
+		rz_il_bv_free(tmp);
+		return ret;
+	}
+
+	return NULL; // something wrong
+}
+
+/**
+ * Get the most significant bit of bitvector
+ * \param bv RzILBitVector, operand
+ * \return b bit, bool value of MSB
+ */
+RZ_API bool rz_il_bv_msb(RzILBitVector bv) {
+	return rz_il_bv_get(bv, bv->len - 1);
+}
+
+/**
+ * Get the least significant bit of bitvector
+ * \param bv RzILBitVector, operand
+ * \return b bit, bool value of LSB
+ */
+RZ_API bool rz_il_bv_lsb(RzILBitVector bv) {
+	return rz_il_bv_get(bv, 0);
+}
+
+char *bv_to_string(RzILBitVector bv) {
+	char *ret = (char *)malloc(sizeof(char) * bv->len);
+	for (int i = 0; i < bv->len; ++i) {
+		ret[i] = rz_il_bv_get(bv, i) ? '0' : '1';
+	}
+	return ret;
+}
+
+/**
+ * Print bitvector, debug function
+ * \param bv RzILBitVector, pointer to bitvector
+ */
+RZ_API void rz_il_print_bv(RzILBitVector bv) {
```

Comment:
Move it to the `librz/core/cil.c` instead. Make `rz_il_bv_as_string()` an API instead.